### PR TITLE
SIMPLY-3425 Migrate R2 Dependencies for Adobe DRM (ePubs)

### PR DESCRIPTION
**Pull request by vladimirfedorov**
**What's this do?**
Adds Adobe DRM protection support to R2 with Content Protection

**Why are we doing this? (w/ JIRA link if applicable)**
https://jira.nypl.org/browse/SIMPLY-3425

**How should this be tested? / Do these changes have associated tests?**
Open NYPL, open any Adobe DRM-protected book

**Dependencies for merging? Releasing to production?**
No

**Does this include changes that require a new SimplyE build for QA?**
No

**Has the application documentation been updated for these changes?**
Yes

**Did someone actually run this code to verify it works?**
@vladimirfedorov

### CODE REVIEW

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+#if FEATURE_DRM_CONNECTOR
+
+class AdobeDRMContentProtection: ContentProtection {
+  func open(asset: PublicationAsset, fetcher: Fetcher, credentials: String?, allowUserInteraction: Bool, sender: Any?, completion: @escaping (CancellableResult<ProtectedAsset?, Publication.OpeningError>) -> Void) {
+    
+    // Publication asset must be a FileAsset, as we are opening a file
+    // If not, we can't open and use Adobe DRM
+    guard let fileAsset = asset as? FileAsset else {
+      completion(.failure(.unavailable(nil)))
```
Comment (ettore) [inline Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:20]: should we add a `Log.error` line to this?

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      return
+    }
+    
+    do {
+      // META-INF is a part of .epub structure
+      let encryptionPath = "META-INF/encryption.xml"
```
Comment (ettore) [inline Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:26]: do you have docs/specs that describe this encryption.xml path? It would be good to reference them here.

File: Simplified/Reader2/AdobeDRM/AdobeDRMFetcher.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    self.fetcher = fetcher
+    self.links = fetcher.links
+  }
+  
+  /// Known resources available in the medium, such as file paths on the file system.
+  var links: [Link]
+  
+  /// Get resource such as content file by its link.
+  ///
+  /// `AdobeDRMFetcher` `get` function open .epub resouorces using the `fetcher` passed to `init`.
```
Comment (ettore) [inline Simplified/Reader2/AdobeDRM/AdobeDRMFetcher.swift:40]: typos (resouorces and AAdobe)

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      let protectedAsset = ProtectedAsset(
+        asset: asset,
+        fetcher: adobeFetcher,
+        onCreatePublication: nil
+      )
+      completion(.success(protectedAsset))
+    } catch {
+      // .success(nil) means it is not an asset protected with Adobe DRM
+      // Streamer continues to iterate over available ContentProtections
+      completion(.success(nil))
```
Comment (ettore) [inline Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:39]: so this means that no matter the reason why `resource.read().get()` throws, here we are saying "there's no valid Adobe DRM, but there IS an encrypted asset that someone else should decrypt" ?  And basically "Failure" means "there's no encrypted asset" ?

**Review by ettore — COMMENTED** [review]
I have a couple questions but this looks good to me.

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift (hunk 1 above)
Comment (vladimirfedorov) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:20]: It's a good question if it can be anything other than file... FileAsset is the only class implementing PublicationAsset protocol at the moment.

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift (hunk 2 above)
Comment (vladimirfedorov) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:26]: it's in ADEPT code, there's no good documentation about this file, except it must be present in .epub for DRM to work. File data (it describes encryption algorithms used to encrypt each file) is passed to Adobe DRM decryption functions.

File: Simplified/Reader2/AdobeDRM/AdobeDRMFetcher.swift (hunk 3 above)
Comment (vladimirfedorov) [reply Simplified/Reader2/AdobeDRM/AdobeDRMFetcher.swift]: my keyboard is killing me. fixed

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift (hunk 1 above)
Comment (ettore) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:20]: in that case it might be good to add one for future proofing

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift (hunk 4 above)
Comment (vladimirfedorov) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:39]: not exactly. this function is called once, when we open a book. at this moment Streamer doesn't know what Content Protection can work with it, and iterates over Content Protections it has (the app initialise Streamer passing an array of Content Protection classes). If the code in do..catch fails, it means it couldn't initialise AdobeDRMFetcher, but there may be other types of Content Protection that work with it. If we return `.failure` here, Streamer stops iterating and shows the error we pass. It means if the book is not protected, Streamer will show the error and won't try to open it without protection. When Streamer receives `.success` with `nil` value for `ProtectedAsset`, it goes to the next Content Protection.

In short, `.failure` means "this is an adobe DRM protected item, but we can't decrypt it for some reason". `catch` block is called when we have no DRM at all.

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift (hunk 1 above)
Comment (vladimirfedorov) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:21]: added

File: Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift (hunk 4 above)
Comment (ErnestFan) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:40]: Would `resource.read().get()` comes back with valid data if the resource is LCP ePub?
If that's the case, we will not be calling `completion(nil)`

Comment (vladimirfedorov) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:40]: > Would resource.read().get() comes back with valid data if the resource is LCP ePub?

No, LCP `.epub`s don't contain `META-INF/encryption.xml`. LCP checks for `.lcpl` in `.epub` (which is embedded into the book during fulfillment process).

Comment (ettore) [reply Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift:40]: that makes sense. could you add that description to the code? it’s makes things much clearer.

**Review by ettore — APPROVED** [review]

**Review by ErnestFan — APPROVED** [review]
